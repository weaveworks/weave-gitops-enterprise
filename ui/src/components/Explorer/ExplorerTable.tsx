import { Box } from '@material-ui/core';
import {
  Flex,
  Icon,
  IconType,
  Link,
  V2Routes,
  formatURL,
} from '@weaveworks/weave-gitops';
import styled from 'styled-components';
import { Object } from '../../api/query/query.pb';
import { Routes, getKindRoute } from '../../utils/nav';
// @ts-ignore
import { DataTable } from '@weaveworks/weave-gitops';
import { Field } from '@weaveworks/weave-gitops/ui/components/DataTable';
import _ from 'lodash';
import { QueryState } from './hooks';

export type FieldWithIndex = Field & { index?: number };

type Props = {
  className?: string;
  onColumnHeaderClick?: (field: Field) => void;
  rows: Object[];
  queryState: QueryState;
  enableBatchSync?: boolean;
  sortField?: string;
  extraColumns?: FieldWithIndex[];
};

function ExplorerTable({
  className,
  rows,
  enableBatchSync,
  sortField,
  onColumnHeaderClick,
  extraColumns = [],
}: Props) {
  const r: Object[] = _.map(rows, o => ({
    // Doing some things here to make this work with the DataTable.
    // It handles rendering the sync/pause buttons.
    ...o,
    uid: o.id,
    clusterName: o.cluster,
    type: o.kind,
  }));

  const fields: FieldWithIndex[] = [
    {
      label: 'Name',
      value: (o: Object) => {
        const page = getKindRoute(o?.kind as string);

        let url: string;
        if (page === V2Routes.NotImplemented) {
          url = formatURL(Routes.ExplorerView, {
            kind: o.kind,
            name: o.name,
            namespace: o.namespace,
            clusterName: o.cluster,
          });
        } else if (page == Routes.Templates){
          url = formatURL(page, {
            name: o.name,
            namespace: o.namespace,
          });
        } else {
          url = formatURL(page, {
            name: o.name,
            namespace: o.namespace,
            clusterName: o.cluster,
          });
        }

        return <Link to={url}>{o.name}</Link>;
      },
      sortValue: () => 'name',
      defaultSort: sortField === 'name',
    },
    { label: 'Kind', value: 'kind', defaultSort: sortField === 'kind' },
    {
      label: 'Namespace',
      value: 'namespace',
      defaultSort: sortField === 'namespace',
    },
    {
      label: 'Cluster',
      value: 'clusterName',
      defaultSort: sortField === 'clusterName',
    },
    {
      label: 'Tenant',
      value: 'tenant',
      defaultSort: sortField === 'tenant',
    },
    {
      label: 'Status',
      sortValue: () => 'status',
      defaultSort: sortField === 'status',
      value: (o: Object) => {
        if (o.status === '-') {
          return '-';
        }

        return (
          <Flex align>
            <Box marginRight={1}>
              <Icon
                size={24}
                color={
                  o?.status === 'Success' ? 'successOriginal' : 'alertOriginal'
                }
                type={
                  o?.status === 'Success'
                    ? IconType.SuccessIcon
                    : IconType.ErrorIcon
                }
              />
            </Box>

            {o?.status}
          </Flex>
        );
      },
    },
    {
      label: 'Message',
      value: 'message',
      defaultSort: sortField === 'message',
    },
  ];

  // Allows for columns to be added anywhere in the table.
  // We sort them here because if we mutate fields out of order,
  // the column order won't be accurate. See test case for example.
  for (const extra of _.sortBy(extraColumns, 'index')) {
    if (typeof extra.index !== 'undefined') {
      fields.splice(extra.index, 0, extra);
    } else {
      fields.push(extra);
    }
  }

  return (
    <DataTable
      className={className}
      fields={fields}
      rows={r}
      hideSearchAndFilters
      onColumnHeaderClick={onColumnHeaderClick}
      hasCheckboxes={enableBatchSync}
      disableSort
    />
  );
}

export default styled(ExplorerTable).attrs({ className: ExplorerTable.name })`
  width: 100%;

  td:nth-child(6),
  td:nth-child(7) {
    white-space: pre-wrap;
    overflow-wrap: break-word;
    word-wrap: break-word;
  }

  /* Moving the sync/pause buttons to the left */
  & > div:first-child {
    justify-content: flex-start;
  }
`;
