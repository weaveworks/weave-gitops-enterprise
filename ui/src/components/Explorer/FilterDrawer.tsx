import { Box } from '@material-ui/core';
import * as React from 'react';
import styled from 'styled-components';

type Props = {
  className?: string;
  children?: React.ReactNode;
  open?: boolean;
  onClose?: () => void;
};

const SlideContainer = styled.div`
  position: relative;
  height: 100%;
  transition-property: width, left;
  transition-duration: 0.5s;
  transition-timing-function: linear;
  overflow: hidden;
  width: 0px;
  flex: 2;
  left: 24px;

  &.open {
    left: 0;
    width: 360px;
  }
`;

const SlideContent = styled.div`
  height: 100%;
  width: 100%;
  border-left: 2px solid ${props => props.theme.colors.neutral20};
  padding-left: ${props => props.theme.spacing.large};
`;

function FilterDrawer({ className, children, open, onClose }: Props) {
  return (
    <div className={className}>
      <SlideContainer className={open ? 'open' : ''}>
        <SlideContent>
          <Box p={1}>
            <div>
              <h2>Filters</h2>
            </div>
            <div>{children}</div>
          </Box>
        </SlideContent>
      </SlideContainer>
    </div>
  );
}

export default styled(FilterDrawer).attrs({ className: FilterDrawer.name })`
  overflow: hidden;
  white-space: nowrap;
`;
