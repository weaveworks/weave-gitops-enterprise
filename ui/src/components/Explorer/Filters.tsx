import { Checkbox, FormControlLabel } from '@material-ui/core';
import _ from 'lodash';
import styled from 'styled-components';
import { Facet } from '../../api/query/query.pb';
import { useReadQueryState, useSetQueryState } from './hooks';

type Props = {
  className?: string;

  facets?: Facet[];
};

function Filters({ className, facets }: Props) {
  const queryState = useReadQueryState();
  const setState = useSetQueryState();

  const handleFilterChange = (field: string, key: string, value: boolean) => {
    const record = `${field}:${key}`;

    const existing = _.find(queryState.filters, f => f === record);

    // Reset the offset when filters change.
    const offset = 0;

    const filters =
      existing && !value
        ? _.filter(queryState.filters, f => f !== record)
        : _.concat(queryState.filters, record);

    setState({
      ...queryState,
      filters,
      offset,
    });
  };

  const filterState = _.reduce(
    queryState.filters,
    (result, f) => {
      const re = /(.+?):(.*)/g;

      const matches = re.exec(f);

      if (matches) {
        const [, key, value] = matches;

        result[`${key.replace('+', '')}:${value}`] = true;
      }

      return result;
    },
    {} as { [key: string]: boolean },
  );

  return (
    <div className={className}>
      {_.map(facets, f => {
        return (
          <div key={f.field}>
            <h3>{_.capitalize(f.field)}</h3>
            <ul style={{ listStyle: 'none' }}>
              {_.map(f.values, v => {
                const key = `${f.field}:${v}`;

                return (
                  <li key={v}>
                    <FormControlLabel
                      label={v}
                      control={
                        <Checkbox
                          // Leaving this as uncontrolled for now.
                          // URL state is proving to be a problem and
                          // the PR is already very sprawling.
                          checked={filterState[key] || false}
                          onChange={e => {
                            handleFilterChange(
                              f.field as string,
                              v,
                              e.target.checked,
                            );
                          }}
                        />
                      }
                    />
                  </li>
                );
              })}
            </ul>
          </div>
        );
      })}
    </div>
  );
}

export default styled(Filters).attrs({ className: Filters.name })`
  ul {
    padding: 0px 8px;
  }

  h2 {
    margin-top: 0;
  }

  h3 {
    margin-bottom: 0;
  }
`;
