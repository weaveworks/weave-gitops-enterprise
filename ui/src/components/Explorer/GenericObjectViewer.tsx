import { Object } from '../../api/query/query.pb';
import { useQueryService } from '../../hooks/query';
import CodeView from '../CodeView';
import KeyValueTable from '../KeyValueTable';
import { LoadingPage } from '@weaveworks/weave-gitops';
import _ from 'lodash';
import styled from 'styled-components';

export type ObjectViewerProps = {
  className?: string;
  kind?: string;
  name?: string;
  namespace?: string;
  clusterName?: string;
};

function GenericObjectViewer({ className, ...rest }: ObjectViewerProps) {
  const { data, isLoading } = useQueryService({
    filters: [],
  });

  if (isLoading) {
    return <LoadingPage />;
  }

  const filters = _.pick(rest, ['kind', 'name', 'namespace']) as any;
  //   Correcting for a inconsistency with the clusterName vs cluster field.
  filters.cluster = rest.clusterName;

  const relevant = _.filter(data?.objects, filters) as Object[];
  const obj = _.first(relevant);

  return (
    <div className={className}>
      <KeyValueTable
        pairs={[
          ['Name', rest.name],
          ['Namespace', rest.namespace],
          ['Kind', rest.kind],
          ['Cluster', rest.clusterName],
          ['Status', obj?.status],
          ['Message', obj?.message],
        ]}
      />
      <CodeView
        code={JSON.stringify(JSON.parse(obj?.unstructured || '{}'), null, 2)}
      />
    </div>
  );
}

export default styled(GenericObjectViewer).attrs({
  className: GenericObjectViewer.name,
})``;
