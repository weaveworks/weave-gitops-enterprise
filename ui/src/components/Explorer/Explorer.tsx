// @ts-ignore
import { Facet } from '../../api/query/query.pb';
import { useListFacets, useQueryService } from '../../hooks/query';
import ExplorerTable, { FieldWithIndex } from './ExplorerTable';
import FilterDrawer from './FilterDrawer';
import Filters from './Filters';
import {
  QueryStateProvider,
  columnHeaderHandler,
  useGetUnstructuredObjects,
} from './hooks';
import PaginationControls from './PaginationControls';
import QueryInput from './QueryInput';
import QueryStateChips from './QueryStateChips';
import { QueryStateManager, URLQueryStateManager } from './QueryStateManager';
import { IconButton } from '@material-ui/core';
import { Alert } from '@material-ui/lab';
import { Flex, Icon, IconType } from '@weaveworks/weave-gitops';
import _ from 'lodash';
import { useState } from 'react';
import { useHistory } from 'react-router-dom';
import styled from 'styled-components';

type Props = {
  className?: string;
  category?: 'automation' | 'source' | 'gitopsset' | 'template';
  enableBatchSync?: boolean;
  manager?: QueryStateManager;
  extraColumns?: FieldWithIndex[];
  linkToObject?: boolean;
};

function Explorer({
  className,
  category,
  enableBatchSync,
  manager,
  extraColumns,
  linkToObject,
}: Props) {
  const history = useHistory();
  if (!manager) {
    manager = new URLQueryStateManager(history);
  }

  const [filterDrawerOpen, setFilterDrawerOpen] = useState(false);
  const { data: facetsRes } = useListFacets();
  const queryState = manager.read();
  const setQueryState = manager.write;

  const { data, error } = useQueryService({
    terms: queryState.terms,
    filters: queryState.filters,
    limit: queryState.limit,
    offset: queryState.offset,
    orderBy: queryState.orderBy,
    ascending: queryState.orderAscending,
    category,
  });

  const unst = useGetUnstructuredObjects(data?.objects || []);
  const rows = _.map(data?.objects, (o: any) => ({
    ...o,
    parsed: unst[o.id],
  }));

  const filteredFacets = filterFacetsForCategory(facetsRes?.facets, category);

  return (
    <QueryStateProvider manager={manager}>
      <div className={className}>
        {error && <Alert severity="error">{error.message}</Alert>}
        <Flex align wide end>
          <QueryStateChips />
          <IconButton onClick={() => setFilterDrawerOpen(!filterDrawerOpen)}>
            <Icon size="normal" type={IconType.FilterIcon} color="neutral30" />
          </IconButton>
        </Flex>
        <Flex wide>
          <ExplorerTable
            queryState={queryState}
            rows={rows}
            onColumnHeaderClick={columnHeaderHandler(queryState, setQueryState)}
            enableBatchSync={enableBatchSync}
            sortField={queryState.orderBy}
            extraColumns={extraColumns}
            linkToObject={linkToObject}
          />

          <FilterDrawer
            onClose={() => setFilterDrawerOpen(false)}
            open={filterDrawerOpen}
          >
            <QueryInput />

            <Filters facets={filteredFacets || []} />
          </FilterDrawer>
        </Flex>

        <PaginationControls
          queryState={queryState}
          setQueryState={setQueryState}
          count={data?.objects?.length || 0}
        />
      </div>
    </QueryStateProvider>
  );
}

export default styled(Explorer).attrs({ className: Explorer.name })`
  width: 100%;
`;

const categoryKinds = {
  automation: ['Kustomization', 'HelmRelease'],
  source: [
    'GitRepository',
    'HelmRepository',
    'Bucket',
    'HelmChart',
    'OCIRepository',
  ],
  gitopsset: ['GitOpsSet'],
  template: ['Template'],
};

function filterFacetsForCategory(
  facets?: Facet[],
  category?: 'automation' | 'source' | 'gitopsset' | 'template',
): Facet[] {
  if (!category) {
    return _.sortBy(facets, 'field') as Facet[];
  }

  const withoutKind = _.filter(facets, facet => {
    return facet.field?.toLowerCase() !== 'kind';
  });

  const kindFacets = _.map(categoryKinds[category], k => k);

  withoutKind.unshift({
    field: 'kind',
    values: kindFacets,
  });

  return _.sortBy(withoutKind, 'field');
}
