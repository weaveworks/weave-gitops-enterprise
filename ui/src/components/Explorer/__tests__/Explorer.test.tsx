import {
  act,
  queryByLabelText,
  render,
  RenderResult,
} from '@testing-library/react';
import {
  defaultContexts,
  MockQueryService,
  withContext,
} from '../../../utils/test-utils';
import Explorer from '../Explorer';
import { addFieldsWithIndex } from '../ExplorerTable';
import { APIContext } from '../../../contexts/API';

describe('addExplorerFields', () => {
  const newField = (id: string, index?: number) => ({
    id,
    label: id,
    value: id,
    index,
  });

  it('adds items to the end of the list if no index', () => {
    const fields = [newField('foo')];
    const fieldsToAdd = [newField('bar')];
    const newFields = addFieldsWithIndex(fields, fieldsToAdd);

    expect(newFields.map(field => field.id)).toEqual(['foo', 'bar']);
  });

  it('adds items to the beginning of the list if index is 0', () => {
    const fields = [newField('foo')];
    const fieldsToAdd = [newField('bar', 0)];
    const newFields = addFieldsWithIndex(fields, fieldsToAdd);
    expect(newFields.map(field => field.id)).toEqual(['bar', 'foo']);
  });

  it('adds item to middle of list if index is specified', () => {
    const fields = [newField('foo'), newField('baz')];
    const fieldsToAdd = [newField('bar', 1)];
    const newFields = addFieldsWithIndex(fields, fieldsToAdd);
    expect(newFields.map(field => field.id)).toEqual(['foo', 'bar', 'baz']);
  });
});

describe('Explorer', () => {
  let wrap: (el: JSX.Element) => JSX.Element;
  let api: MockQueryService;

  beforeEach(() => {
    api = new MockQueryService();
    wrap = withContext([
      ...defaultContexts(),
      [APIContext.Provider, { value: { query: api } }],
    ]);
  });

  it('renders rows', async () => {
    const objects = [
      {
        kind: 'Kustomization',
        name: 'flux-system',
        namespace: 'flux-system',
        status: 'Ready',
      },
      {
        kind: 'HelmRelease',
        name: 'flux-system',
        namespace: 'flux-system',
        status: 'Ready',
      },
    ];

    api.DoQueryReturns = {
      objects,
    };

    let result: RenderResult;
    await act(async () => {
      const c = wrap(<Explorer />);
      result = await render(c);
    });

    // @ts-ignore
    expect(result.container).toHaveTextContent(objects[0].name);
  });
  it('renders filter elements', async () => {
    api.DoQueryReturns = {
      objects: [],
    };

    const facets = [
      {
        field: 'Kind',
        values: ['Kustomization', 'HelmRelease', 'GitRepository'],
      },
    ];

    api.ListFacetsReturns = {
      facets,
    };

    let result: RenderResult;

    await act(async () => {
      const c = wrap(<Explorer />);
      result = await render(c);
    });

    // @ts-ignore
    expect(result.container).toHaveTextContent(facets[0].field);
    // @ts-ignore
    expect(result.container).toHaveTextContent(facets[0].values[0]);
  });
  it('renders filter state', async () => {
    api.DoQueryReturns = {
      objects: [],
    };

    const facets = [
      {
        field: 'Kind',
        values: ['Kustomization', 'HelmRelease', 'GitRepository'],
      },
    ];

    api.ListFacetsReturns = {
      facets,
    };

    const manager = {
      read: jest.fn(() => ({
        terms: '',
        filters: ['Kind:Kustomization'],
        limit: 0,
        offset: 0,
        orderBy: '',
        orderDescending: false,
      })),

      write: jest.fn(),
    };

    let result: RenderResult;

    await act(async () => {
      const c = wrap(<Explorer manager={manager} />);
      result = await render(c);
    });

    //  @ts-ignore
    const container = result.container;
    //  @eslint-disable-next-line
    const input = queryByLabelText(
      container,
      'Kustomization',
    ) as HTMLInputElement;

    expect(input?.checked).toBeTruthy();
  });
  it('you can configure the visible columns', async () => {
    const objects = [
      {
        kind: 'Kustomization',
        name: 'flux-system',
        namespace: 'flux-system',
        status: 'Ready',
      },
      {
        kind: 'HelmRelease',
        name: 'flux-system',
        namespace: 'flux-system',
        status: 'Ready',
      },
    ];

    api.DoQueryReturns = {
      objects,
    };

    const tableFields = [
      {
        id: 'my-cool-column',
        label: 'My Cool Column',
        value: (o: any) => `${o.kind}-foo-bar`,
      },
    ];

    let result = {} as RenderResult;
    await act(async () => {
      const c = wrap(<Explorer fields={tableFields} />);
      result = await render(c);
    });

    expect(result.container).toHaveTextContent('My Cool Column');
    expect(result.container).toHaveTextContent('Kustomization-foo-bar');
  });

  describe('snapshots', () => {
    it('renders', async () => {
      let result: RenderResult;

      api.DoQueryReturns = {
        objects: [
          {
            kind: 'Kustomization',
            name: 'flux-system',
            namespace: 'flux-system',
            status: 'Ready',
          },
        ],
      };
      api.ListFacetsReturns = {
        facets: [
          {
            field: 'Kind',
            values: ['Kustomization', 'HelmRelease', 'GitRepository'],
          },
        ],
      };

      await act(async () => {
        const c = wrap(<Explorer />);
        result = await render(c);
      });

      //   @ts-ignore
      expect(result.container).toMatchSnapshot();
    });
  });
});
