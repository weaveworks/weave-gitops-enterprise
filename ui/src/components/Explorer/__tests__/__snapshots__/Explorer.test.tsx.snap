// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explorer snapshots renders 1`] = `
<div>
  <div
    class="cWzBAr Explorer"
  >
    <div
      class="cSydOs"
    >
      <div
        class="QueryStateChips"
      >
        <div
          class="eqkCsf cTdykF $9121bed0097d59fd$var$ChipGroup"
        />
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <div
            class="fNCJgN gcZbac"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
              />
            </svg>
          </div>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="kWyVmm"
    >
      <div
        class="gNvEgP dpBfnz iWRZYi ExplorerTable"
      >
        <div
          class=""
        />
        <div
          class="cSydOs fGHQuv"
        />
        <div
          class="fXhDuD"
        >
          <div
            class="MuiTableContainer-root"
          >
            <table
              aria-label="simple table"
              class="MuiTable-root"
            >
              <thead
                class="MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    scope="col"
                  >
                    <div
                      class="IsKSv"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text lfZVQa MuiButton-colorInherit"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          <h2
                            class="selected"
                          >
                            Name
                          </h2>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                      <div
                        class="fmllGp"
                      />
                      <div
                        class="fNCJgN hxXmRH downward"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                          />
                        </svg>
                      </div>
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    scope="col"
                  >
                    <div
                      class="IsKSv"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text lfZVQa MuiButton-colorInherit"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          <h2
                            class=""
                          >
                            Kind
                          </h2>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                      <div
                        class="fmllGp"
                      />
                      <div
                        style="width: 16px;"
                      />
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    scope="col"
                  >
                    <div
                      class="IsKSv"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text lfZVQa MuiButton-colorInherit"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          <h2
                            class=""
                          >
                            Namespace
                          </h2>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                      <div
                        class="fmllGp"
                      />
                      <div
                        style="width: 16px;"
                      />
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    scope="col"
                  >
                    <div
                      class="IsKSv"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text lfZVQa MuiButton-colorInherit"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          <h2
                            class=""
                          >
                            Cluster
                          </h2>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                      <div
                        class="fmllGp"
                      />
                      <div
                        style="width: 16px;"
                      />
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    scope="col"
                  >
                    <div
                      class="IsKSv"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text lfZVQa MuiButton-colorInherit"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          <h2
                            class=""
                          >
                            Tenant
                          </h2>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                      <div
                        class="fmllGp"
                      />
                      <div
                        style="width: 16px;"
                      />
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    scope="col"
                  >
                    <div
                      class="IsKSv"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text lfZVQa MuiButton-colorInherit"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          <h2
                            class=""
                          >
                            Status
                          </h2>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                      <div
                        class="fmllGp"
                      />
                      <div
                        style="width: 16px;"
                      />
                    </div>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head"
                    scope="col"
                  >
                    <div
                      class="IsKSv"
                    >
                      <button
                        class="MuiButtonBase-root MuiButton-root MuiButton-text lfZVQa MuiButton-colorInherit"
                        tabindex="0"
                        type="button"
                      >
                        <span
                          class="MuiButton-label"
                        >
                          <h2
                            class=""
                          >
                            Message
                          </h2>
                        </span>
                        <span
                          class="MuiTouchRipple-root"
                        />
                      </button>
                      <div
                        class="fmllGp"
                      />
                      <div
                        style="width: 16px;"
                      />
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    <span
                      class="jNLgwU"
                    >
                      <a
                        class="fOLcns $8405a4f7caf64b1d$var$Link"
                        href="/kustomization?name=flux-system&namespace=flux-system"
                      >
                        <span
                          class="ginmAQ"
                          color="primary"
                        >
                          flux-system
                        </span>
                      </a>
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    <span
                      class="jNLgwU"
                    >
                      Kustomization
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    <span
                      class="jNLgwU"
                    >
                      flux-system
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    <span
                      class="jNLgwU"
                    >
                      -
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    <span
                      class="jNLgwU"
                    >
                      -
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    <span
                      class="jNLgwU"
                    >
                      <div
                        class="fNCJgN"
                      >
                        <div
                          class="MuiBox-root MuiBox-root-31"
                        >
                          <div
                            class="fNCJgN dGvMZw"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
                              />
                            </svg>
                          </div>
                        </div>
                        Ready
                      </div>
                    </span>
                  </td>
                  <td
                    class="MuiTableCell-root MuiTableCell-body"
                  >
                    <span
                      class="jNLgwU"
                    >
                      -
                    </span>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
      <div
        class="ljFUpk FilterDrawer"
      >
        <div
          class="kfgeSP"
        >
          <div
            class="gMuAtv"
          >
            <div
              class="MuiBox-root MuiBox-root-29"
            >
              <div>
                <h2>
                  Filters
                </h2>
              </div>
              <div>
                <div
                  class="kWyVmm cUaLCy QueryInput"
                >
                  <div
                    class="fNCJgN"
                  >
                    <div
                      class="fNCJgN juDKMA"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                        />
                      </svg>
                    </div>
                    <div
                      class="MuiFormControl-root"
                    >
                      <div
                        class="MuiFormControl-root MuiTextField-root $66a683f19021b9f2$var$Input"
                      >
                        <div
                          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                        >
                          <input
                            aria-invalid="false"
                            class="MuiInputBase-input MuiInput-input"
                            placeholder="Search"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="dZCiSF Filters"
                >
                  <div>
                    <h3>
                      Kind
                    </h3>
                    <ul
                      style="list-style: none;"
                    >
                      <li>
                        <label
                          class="MuiFormControlLabel-root"
                        >
                          <span
                            aria-disabled="false"
                            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-32 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                          >
                            <span
                              class="MuiIconButton-label"
                            >
                              <input
                                class="PrivateSwitchBase-input-35"
                                data-indeterminate="false"
                                type="checkbox"
                                value=""
                              />
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                                />
                              </svg>
                            </span>
                            <span
                              class="MuiTouchRipple-root"
                            />
                          </span>
                          <span
                            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
                          >
                            Kustomization
                          </span>
                        </label>
                      </li>
                      <li>
                        <label
                          class="MuiFormControlLabel-root"
                        >
                          <span
                            aria-disabled="false"
                            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-32 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                          >
                            <span
                              class="MuiIconButton-label"
                            >
                              <input
                                class="PrivateSwitchBase-input-35"
                                data-indeterminate="false"
                                type="checkbox"
                                value=""
                              />
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                                />
                              </svg>
                            </span>
                            <span
                              class="MuiTouchRipple-root"
                            />
                          </span>
                          <span
                            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
                          >
                            HelmRelease
                          </span>
                        </label>
                      </li>
                      <li>
                        <label
                          class="MuiFormControlLabel-root"
                        >
                          <span
                            aria-disabled="false"
                            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-32 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                          >
                            <span
                              class="MuiIconButton-label"
                            >
                              <input
                                class="PrivateSwitchBase-input-35"
                                data-indeterminate="false"
                                type="checkbox"
                                value=""
                              />
                              <svg
                                aria-hidden="true"
                                class="MuiSvgIcon-root"
                                focusable="false"
                                viewBox="0 0 24 24"
                              >
                                <path
                                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                                />
                              </svg>
                            </span>
                            <span
                              class="MuiTouchRipple-root"
                            />
                          </span>
                          <span
                            class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
                          >
                            GitRepository
                          </span>
                        </label>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="jfFvvd PaginationControls"
    >
      <div
        class="MuiBox-root MuiBox-root-30"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <div
              class="fNCJgN juDKMA"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </div>
          </span>
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <div
              class="fNCJgN juDKMA"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </div>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
